### ISSUE TITLE
[1.8.2] Development environment build time increased 150%

### ISSUE THREAD

**Issue by theosp** [meteor/meteor#10800]
I am in the process of evaluating v1.8.2 for [JustDo.com](https://justdo.com) (upgrading from v1.8.1).

Development app start time takes 150% longer - from 20 seconds, it now increased to 50 seconds (2.9 GHz Intel Core i9, MBP 2018).

JustDo is comprised of 2 apps, one for the landing app and one for the web app itself.

The following is the output received when running the script that runs the two apps with the env var `METEOR_PROFILE=1`.

[JustDo v1.8.1 profile](https://drive.google.com/open?id=1HRtzhn1E2HRIS-cWFGurLzfQbu7srKDv)
[JustDo v1.8.2 profile](https://drive.google.com/open?id=1gkIM6ljW_6Czrif-Lopv9zttEQ9XteiZ)

The output isn't for the first run, but for the second run of the apps in both cases (to avoid special inits when changing meteor versions).

Notes:

* Setting env var `METEOR_DISABLE_OPTIMISTIC_CACHING=1` didn't help
* https://github.com/meteor/meteor/issues/10786 - Might be relevant?

**Comment by sebakerckhof** [comment]
Are you sure these have been created in similar conditions? The fact that some things like reading in the SQLite package database etc all take considerably longer would suggest there might have been some heavy resource usage (either disk/memory or cpu) by something unrelated?

Since a lot of time is spend by the linker, could you also try running with:
`METEOR_TEST_PRINT_LINKER_CACHE_DEBUG=1` envar

**Comment by theosp** [comment]
Hi thanks for the reply,

1. In the development environment we are not connecting to an SQL server, nor any other similar resources that can have similar effect on results.
2. This time, again, the output is from the *2nd* run of the environment, and not from the first run.
3. The outputs are from the exact same code, as the original output, only with the additional env var: METEOR_TEST_PRINT_LINKER_CACHE_DEBUG=1
4. To speed the process I published the results bottom line under each link.

I ran the test twice for each version, one with METEOR_DISABLE_OPTIMISTIC_CACHING=1 and one without.

[JustDo v1.8.1 profile](https://drive.google.com/open?id=1H5IVy7lPNMpxbj35mDn2bWkvc20c5zmB)
Total: 14,950 ms (Build App) (Web app)
Total: 6,045 ms (Build App) (Landing  app)

[JustDo v1.8.1 profile-METEOR_DISABLE_OPTIMISTIC_CACHING=1](https://drive.google.com/open?id=1iWDP9BCFzabQL1YDVu9QhTImJ3Ve88Np)
Total: 10,235 ms (Build App) (Web app)
Total: 4,699 ms (Build App) (Landing  app)

[JustDo v1.8.2 profile](https://drive.google.com/open?id=1vlsqHbGruRz9gOcJqIflW5Nw7ZT-_i0o)
Total: 39,903 ms (Build App) (Web app)
Total: 11,746 ms (Build App) (Landing  app)

[JustDo v1.8.2 profile-METEOR_DISABLE_OPTIMISTIC_CACHING=1](https://drive.google.com/open?id=1F42kVjAaws5OtFpqnG3YGXVjIlCcmzYJ)
Total: 31,009 ms (Build App) (Web app)
Total: 9,884 ms (Build App) (Landing  app)

As you can see, I was wrong to say METEOR_DISABLE_OPTIMISTIC_CACHING has no effect, but still v1.8.1 with it builds *3 times* faster compared to v1.8.2!

**Comment by sebakerckhof** [comment]
I'll have a look at the profiles later, but the sql server I'm talking about is the Meteor's package sqlite database.

**Comment by theosp** [comment]
Thank you.

re sqlite, the results are consistent enough to rule out occasional heavy resource usage.

**Comment by filipenevola** [comment]
@theosp are you using dynamic imports? 

I included a fix on Meteor 1.8.2 that makes dynamic import to work in every case because before 1.8.2 dynamic import could fail for [some cases](https://github.com/meteor/meteor/pull/10730).

Maybe you were lucky and the issue was not affecting you but now that Meteor is importing everything as expected this could have increased your build time. I'm not affirming this, it is just a hypothesis.

Also as Meteor 1.9 will be released soon could you try the same using the latest 1.9 beta and report back the numbers?

**Comment by theosp** [comment]
@sebakerckhof , since v1.8.1 already did rebuilds for `'web.browser.legacy'` while your work will probably improve results, it doesn't touch the actual degradation in build time speed between v1.8.1 to v1.8.2 .

Need to look for the actual cause, perhaps `git bisect` can be handy to track down the problematic change.

**Comment by sebakerckhof** [comment]
@theosp Yes, I'm not claiming that PR is a solution for this specific problem. I just referred this issue to show rebuild times are still a hot topic and excluding legacy architectures during development can be one thing to help.

**Comment by benjamn** [comment]
> Need to look for the actual cause, perhaps `git bisect` can be handy to track down the problematic change.

@theosp You may have better/quicker results if you manually bisect the release commits (`-beta.n` and `-rc.n`), rather than using `git bisect`.

Regarding `sqlite3`, I thought we had upgraded it in 1.8.2, which might explain the performance regression, but I was wrong; we only updated it (to 4.1.0, from 3.1.8) in Meteor 1.9. Nevertheless, it does seem worth understanding why reading from the package DB is taking so much longer now.

**Comment by benjamn** [comment]
Another thing that stands out is that `other PackageSourceBatch#_linkJS` is taking longer in 1.8.2, which suggests there's something goin on in [this code](https://github.com/meteor/meteor/blob/release-1.8.2/tools/isobuild/compiler-plugin.js#L1641-L1790) that isn't very well captured by the profiling tool.

**Comment by sebakerckhof** [comment]
@benjamn This comment contains logs with `METEOR_TEST_PRINT_LINKER_CACHE_DEBUG=1`
https://github.com/meteor/meteor/issues/10800#issuecomment-562988826

But if anything, there's more cache misses in the 1.8.1 version, so at least that's not the problem in the linking phase.

@theosp Unfortunately I'm not able to reproduce this with our large meteor app (in fact 1.8.2 seems a little faster for us).

Some things that you still could try or could help identify the problem:
- Could you try running meteor with `TOOL_NODE_FLAGS=--inspect-brk` envar and take a cpu profile with chrome dev tools?

- Can you also try with `TOOL_NODE_FLAGS="--max-old-space-size=6192 --no-expose-gc"` to see if it's maybe memory-trashing issue where the garbage collector takes up a lot of CPU (although I wouldn't expect to already see performance impact in the `prepareProjectForBuild` stage if that's the case)

**Comment by heschong** [comment]
Posted some info in the [Forums](https://forums.meteor.com/t/2x-to-3x-longer-build-time-also-on-testing-since-1-8-2-update/51028/10?u=heschong), should have posted here instead.  In our app, when we save a single change to a single file in one package, the rebuild times skyrocketed under 1.8.2 (and 1.8.3 and 1.9b4).  It seems stuck on that call to sqlite (here with `METEOR_PROFILE=100`):

```
| ProjectContext prepareProjectForBuild.......................331,917 ms (1)
| ├─ _initializeCatalog.......................................330,436 ms (1)
| │  └─ LocalCatalog#initialize...............................330,435 ms (1)
| │     └─ LocalCatalog#_loadLocalPackages....................330,427 ms (1)
| │        └─ PackageSource#initFromPackageDir................330,398 ms (15)
| │           └─ sqlite query                                 330,317 ms (13)
```

but our results with going back down to 1.8.1 are much different (with `METEOR_PROFILE=1`):

```
| ProjectContext prepareProjectForBuild.........................1,234 ms (1)
| ├─ _initializeCatalog...........................................133 ms (1)
| │  └─ LocalCatalog#initialize...................................133 ms (1)
| │     ├─ addPatternsToList                                       41 ms (2)
| │     ├─ LocalCatalog#_computeEffectiveLocalPackages.............23 ms (1)
| │     │  ├─ files.readdir                                         2 ms (2)
| │     │  ├─ files.stat                                            6 ms (13)
| │     │  └─ other LocalCatalog#_computeEffectiveLocalPackages    16 ms
| │     └─ LocalCatalog#_loadLocalPackages.........................68 ms (1)
| │        ├─ PackageSource#initFromPackageDir.....................64 ms (15)
| │        │  ├─ files.stat                                         9 ms (15)
| │        │  ├─ runJavaScript package.js                           3 ms (15)
| │        │  ├─ sqlite query                                      44 ms (13)
| │        │  └─ other PackageSource#initFromPackageDir             7 ms
| │        └─ other LocalCatalog#_loadLocalPackages                 4 ms
```

Every time we touch a file in one package, it completely (and very slowly) rebuilds every other package we have in our `packages/` directory, whereas it seems like in 1.8.1 only the one package is re-examined (or if the others are they go by fast enough to not notice)

**Comment by benjamn** [comment]
@heschong While that's obviously not a great experience, I am very confident that the `sqlite3` performance regression is some sort of fixable bug, rather than a general increase in build complexity with no easy answer.

**Comment by heschong** [comment]
@benjamn

When we turn on `METEOR_DEBUG_BUILD=1` we see that each of our 10 packages takes about 30 seconds, and that's what adds up to the 300 seconds roughly.

One of these 10 packages is imported from most of the other 9.  Which also adds a wrinkle somewhat to the mix.

I should note that these times are from a relatively slow runtime environment (a virtualbox vm) running the app within a docker container that has `.meteor` mounted into the container from the vm (long story)... so disk performance is probably not what we would call great.  But we did run this with `--max-old-space-size=8192` and it was the same setup between versions.  The sqlite comment from @sebakerckhof earlier made me question whether that might actually be a component of an issue.

**Comment by benjamn** [comment]
@heschong I just pushed a commit to `devel` that will improve the `METEOR_PROFILE` output by showing individual package names, though I see you already have that information from `METEOR_DEBUG_BUILD=1`.

Using a SQLite from disk on a VM file system will definitely slow things down. I wonder if there's a way to load the entire database into memory (in a readonly way) to avoid repeatedly accessing the disk? The DB is fairly large (300MB+), so loading the whole thing into memory might not be desirable in all cases.

Further reading for anyone who's interested:
* https://www.sqlite.org/inmemorydb.html
* https://www.sqlite.org/c3ref/open.html

**Comment by heschong** [comment]
@benjamn Cool, we'll check it out.  I think our slow file access to that sqlite db at least amplifies this issue maybe in a way that makes it stick out more.  But the fact that 1.8.1 somehow was able to do what I can only assume is the same task on the same file in the same environment in 44ms vs 330,317ms smells weird.  Will keep digging on my side, thanks for the assistance!

**Comment by heschong** [comment]
@benjamn I found the `METEOR_DEBUG_SQL` setting, turned it on.  In 1.8.3, for each package during that `START 5 reading package` phase, I see stuff like this:

```
SQL statement  BEGIN IMMEDIATE TRANSACTION  took  14449
SQL statement  SELECT content FROM releaseVersions WHERE track=? AND version=?  took  8486
SQL statement  END TRANSACTION  took  15179   
Transaction took:  38119                      
```

But in 1.8.1, I don't see any `SQL statement` debug lines for each package as it's being loaded, just a little line that says:

```
Transaction took: 4
```

It appears that these queries are being executed from this `getReleaseVersion` function:

https://github.com/meteor/meteor/blob/811a56a83f5685c35d9133743b81b53f7ef6597b/tools/packaging/catalog/catalog-remote.js#L666

So...

1) I don't understand why they are being executed in 1.8.3 but not 1.8.1, seems like they should be, as far as I can trace it's execution back.

2) I tried querying the sqlite db file it was reading from during the slow build process, and could not run the `BEGIN IMMEDIATE TRANSACTION` part because it was locked.  Is it taking so long because the sqlite db is locked by something else going on?

3) The result from this query is so small and often repeated that it could easily be memoized or cached probably.


I'm curious if others see a difference in the number of queries and the length when `METEOR_DEBUG_SQL` is on, or if my case is unique and not related to this issue.

**Comment by afrokick** [comment]
@heschong did you try to set `METEOR_SQLITE_JOURNAL_MODE` env to `TRUNCATE`?

https://github.com/meteor/meteor/blob/devel/tools/packaging/catalog/catalog-remote.js#L19-L24

**Comment by heschong** [comment]
@afrokick just tried it out, didn't seem to make a noticeable difference for my environment.  That setting does bring up an interesting point, the only line in that file that materially changed between the 1.8.1 and 1.8.2 release was this one:

https://github.com/meteor/meteor/blob/096641b084b70f4ab52f3ef4468728d164ec66d1/tools/packaging/catalog/catalog-remote.js#L150

```
-    self._execute(`PRAGMA journal_mode=${JOURNAL_MODE}`);
+    self._execute("PRAGMA journal_mode=".concat(JOURNAL_MODE));
```

I can't actually find that second line of code here in GitHub which is, um, unsettling.  But that's the diff when I manually diff those files out of my `.meteor/packages/meteor-tool/<version numbers>` directories.  Anyway that's the literal diff between 1.8.1 and 1.8.2 on my machine. As an experiment, I manually switched that line of code back in my 1.8.2 install and it did not seem to make any impact, so I don't know if this is material to the issue.

**Comment by heschong** [comment]
Another reason I feel like 1.8.2 and above are "loading more code than before" on each rebuild is that after upgrading, we see warnings for npm packages that are not installed, but those npm packages are a) only used server-side, b) not direct dependencies in our app.

Example: meteor above 1.8.1 now says that it's unable to resolve `hiredis` which is not a defined dependency anywhere in our app, but there *is* a reference in a package we have called `redid-parser` that has a `require()` in the code that is inside a try-catch statement, which is inside an if statement. This part of code never gets referenced by our app.

This warning comes up with a handful of npm library child dependencies now, we've just been ignoring it. We've tried installing some of these libraries to make the warning go away, but some are not public or compilable, etc (the `mongo` npm library references `mongo-client-encryption` for instance) and since these are deeply nested in other libraries we can't really remove them.

Anyway my point is, it feels like the build system is pulling in way more code now than before. And I don't know if this sqlite performance stuff is related, or if it's just a byproduct of re-processing or loading extra code.
